src/PHPStan/PregMatchTypeSpecifyingExtension.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\PHPStan;
+
+use PhpCsFixer\Preg;
+use PhpParser\Node\Expr\StaticCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Analyser\SpecifiedTypes;
+use PHPStan\Analyser\TypeSpecifier;
+use PHPStan\Analyser\TypeSpecifierAwareExtension;
+use PHPStan\Analyser\TypeSpecifierContext;
+use PHPStan\Reflection\MethodReflection;
+use PHPStan\TrinaryLogic;
+use PHPStan\Type\Php\RegexArrayShapeMatcher;
+use PHPStan\Type\StaticMethodTypeSpecifyingExtension;
+
+final class PregMatchTypeSpecifyingExtension implements StaticMethodTypeSpecifyingExtension, TypeSpecifierAwareExtension
```

Comment:
i wonder how we can avoid those custom extensions.
Preg::match is proxy to preg_match, can't we rely on php-stan knowledge of preg_match typing?

Reply:
I wonder if we can avoid extension at all and have in-place (close to definition of `Preg::match`) declaration, something like `@returns ReturnType<preg_match>` (if I would do it in typescript - sorry, not that deeply familiar with stan typing)

Reply:
It's not possible. This is the closest we can get and still be flexible. Some `preg_match` "replacements" sometimes have a bit different behaviour in regard to some flags for example. Some replacements still use by-ref parameters, some other replacements return matches instead.